Sequential Channelization Script

hzorkic opened the pull request:
Another smallish PR for a script that channelizes the raw ERA5 or ECMWF data. Mostly validating logic here, but a few words of advice would be useful. 


This script is largely based off of another script Jeramey wrote, but I've modified it to do sequential writes, append data to existing repos, and workwith era5 or ECMWF data. 

## Review 
`channelize.ipynb` : will outline what our input data + output data will look like. 
- notice how there is no metadata ( var_locs ) getting carried over in the final `Channelized ECMWF Data ...` section. I need help figuring out how to make sure that isn't getting dropped. 
- also ISO opinions on any other advice on what other metadata to include here. var_locs should be okay .

`channelize.py` : the core script. 
**- most critical: can you validate that I am correctly appending data to existing stores using sequential_reorg ?**
- again, where is var_locs getting dropped ?
- this is pretty slow ... i know it's partially because i'm doing sequential writes, but i want to see if there are any other hold ups. 
- i know the way i'm loading the source datasets / renaming is a bit clunky. i do think it's OKAY though. 
- how hard would it be to add back in parallelism? I'm specifically weary to do so because of how we're extending the stores for new timetamps. Would i need to move extending the dataset outside of the processing loop. does that make sense?
- I think that the way I'm handling days instead of timestamps isn't the most robust / clear. For example, somebody might want to update a # of timestamps that don't fall exactly into day bins. It might be more clean for the user to input "YYYY-MM-DDTHH:MM:SS" instead of "YYY-MM-DD"

## Don't Review
`dataset.io` : should be sound, feel free to take a gander to better understand the logic in `channelize.py`



## Extras 
`channelize.sh` : a shell script that wraps up the python script so that it'll keep running if there is an exception raised. 

The retry infrastructure here is mostly to address I/O errors from icechunk that looks like this. If there's something specific you can see that might be causing that, please do let me know, but for now, wrapping in a shell script is working fine. 

```
   Batch failure:   x error streaming bytes from object store streaming error
  | 
  | context:
  |    0: icechunk::storage::s3::get_object_range_buf
  |            with settings=Settings { concurrency: None, retries: None, unsafe_use_conditional_update: None, unsafe_use_conditional_create: None, unsafe_use_metadata: None, storage_class: None,
  | metadata_storage_class: None, chunks_storage_class: None, minimum_size_for_multipart_upload: None } key="pjcakaai_model_ecmwf_t0_nonprod_frankfurt/manifests/1R3PZA1D6A8RMC4K44VG"
  | range=20274331..30411496
  |              at icechunk/src/storage/s3.rs:951
  |    1: icechunk::storage::s3::fetch_manifest_known_size
  |            with id=1R3PZA1D6A8RMC4K44VG size=50685826
  |              at icechunk/src/storage/s3.rs:601
  |    2: icechunk::asset_manager::fetch_manifest
  |            with manifest_id=1R3PZA1D6A8RMC4K44VG manifest_size=50685826
  |              at icechunk/src/asset_manager.rs:171
  |    3: icechunk::session::get_chunk_ref
  |            with path=Path(Utf8PathBuf { _encoding: "unix", inner: "/ENFO-T0/v" }) coords=ChunkIndices([9284, 0, 10, 0, 0])
  |              at icechunk/src/session.rs:637
  |    4: icechunk::session::get_chunk_reader
  |            with path=Path(Utf8PathBuf { _encoding: "unix", inner: "/ENFO-T0/v" }) coords=ChunkIndices([9284, 0, 10, 0, 0]) byte_range=From(0)
  |              at icechunk/src/session.rs:701
  |    5: icechunk::store::get
  |            with key="ENFO-T0/v/c/9284/0/10/0/0" byte_range=From(0)
  |              at icechunk/src/store.rs:195
  | 
  |-> error streaming bytes from object store streaming error
  |-> streaming error
  `-> minimum throughput was specified at 1 B/s, but throughput of 0 B/s was observed
 ```

hzorkic committed:
add scripts to channelize era5 or ecmwf data

hzorkic committed:
pre-pr push

jeramey-rwe commented:
I don't think you want to merge this into main. The change set looks huge because of this. Change the PR to merge into `latent-vector-predict`.

hzorkic committed:
new PR update

jeramey-rwe commented on dataset_io.py:
```diff
@@ -0,0 +1,487 @@
+    the general way in which the production data store was initialized.
+
+    Args:
+        store (zarr.abc.store.Store): The Zarr store where the dataset will be saved.
```
You can skip including the type hint in the docstring. It's already easily seen in the function signature.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+
+    idx = 0
+
+    for var in NAME_MAP["sfc"]:
```
Just a word of caution: Python _currently_ preserves dict key insert order, but hasn't done so in the past and may not again in the future. This code should be fine, but always be cautious about assuming the order of keys you'll get out of any mapping object.

hzorkic replied:
noted!

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+
+            commit_msg = f"Added batch {batch[0]:%Y-%m-%d} → {batch[-1]:%Y-%m-%d}"
+
+            # Add retry logic. Sometimes icechunk repo has
```
Incomplete sentence. Icechunk repo has what? I'm also very curious what issue you're running into here where retrying the commit is succeeding. That doesn't make sense to me. If a commit fails once, it should always fail unless its because of a network I/O error. Are you seeing those?

hzorkic replied:
all of my retry logic is indeed because of the I/O error!

jeramey-rwe replied:
Is the failure during `process_day_from_datasets`? If it is, you should constrain the try/except block to just that line.

If you're seeing network I/O failures when you write to a repository local to kafou, I'd like to try to replicate that issue in a small example to diagnose it. That's not something that should happen very often, and certainly should extremely rarely happen at commit time since commits are a very tiny write.

jeramey-rwe replied:
What I'm really trying to get at here from an exception-catching standpoint, is that exception catching and retries should always be constrained to the smallest amount of code possible. If the errors are mostly happening during reads from a public Arraylake repo, you should try/except catch just those lines which perform those reads and retry them with appropriate logging saying what's going on.

If you're _also_ seeing issues with writing, try/except and retry that code _separately_. It'll make diagnosing issues easier and will reduce the amount of work that is done. (E.g. you don't want to have to read from a flaky source again if you had a write failure to a normally robust write destination.)

hzorkic replied:
I moved around the I/O read excetion logic to channelize_day() so the trys are constrained.

hzorkic replied:
the errors are only read errors and they're all streaming errors and they all seem to happen on pressure_level data: 

```
  │    2: icechunk::store::get
  │            with key="pressure_level/U/c/23749/11/2/2" byte_range=From(0)
```

```
× I/O error: streaming error
  │ 
  │ context:
  │    0: icechunk::storage::s3::fetch_chunk
  │            with id=4T8EVD2K8FYQYFC1W33G range=0..358031
  │              at icechunk/src/storage/s3.rs:632
  │    1: icechunk::asset_manager::fetch_chunk
  │            with chunk_id=4T8EVD2K8FYQYFC1W33G range=0..358031
  │              at icechunk/src/asset_manager.rs:288
  │    2: icechunk::store::get
  │            with key="pressure_level/U/c/23749/11/2/2" byte_range=From(0)
  │              at icechunk/src/store.rs:195
  │ 
  ├─▶ I/O error: streaming error
  ├─▶ streaming error
  ╰─▶ minimum throughput was specified at 1 B/s, but throughput of 0 B/s was observed
```

so maybe that data is somehow wacky. To be investigated, but for now the retrys are fine and at least the error handling is more isolated and now, in theory, we could just run channelize_day() on it's own for further investigation.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+                except Exception as e:
+                    print(f"   Commit attempt #{attempt} failed: {e}")
+                    if attempt == 6:
+                        raise RuntimeError("Max commit retries reached.")
```
You probably want to just do `raise` here to re-raise the specific exception so that you get a proper stack trace. You can just use `print()` to print the info about max tries having been reached. Ideally something like

```python
print("Max commit retries reached.", file=sys.stderr, flush=True)
raise
```

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+                    print(f"   COMMIT SUCCESS — {cid}")
+                    last_committed_day = batch[-1].date()
+                    break
+                except Exception as e:
```
In general, try to never catch `Exception`. Always better to catch more specific exception types. You could be suppressing other kinds of exceptions you really shouldn't be doing any retries on and totally hosing something critical in the process.

hzorkic replied:
got it. A recently point of learning for me is better exception logging.

jeramey-rwe replied:
I would also say in addition, try to be very sparing about when you catch exceptions. _Only_ catch them if they're something you can handle and correct for. Otherwise it is better to just let the exception raise and your program crash so that you can get as clean of a stack trace as possible which will help you find and correct problems.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    *,
+    sfc_ds,
+    pl_ds,
+    src_repo_name,
```
This parameter isn't used. Delete it or pass it through to functions which should be using it.

hzorkic replied:
done.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    dst_repo_name,
+    start_time,
+    end_time,
+    src_branch="main",
```
This parameter isn't used. Delete it or pass it through to functions which should be using it.

hzorkic replied:
done.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    # for each batch ...
+    for batch in batches:
+        print(f"\n▶️  BATCH {batch[0]:%Y-%m-%d} → {batch[-1]:%Y-%m-%d}")
+        t0 = time.monotonic()
```
More expressive name would be better. `batch_start_time` maybe?

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+                print("RESTART_CURSOR=NONE")
+                print("   No batches were successfully committed yet.")
+
+            raise SystemExit(1)
```
Don't do this. Just re-raise the exception with a bare `raise`.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    compressors = [
+        zarr.codecs.BloscCodec(
+            clevel=3,
+            shuffle=zarr.codecs.BloscShuffle.bitshuffle,
```
Add a `typesize=4` parameter to this. It will improve compression since the data is np.float32 values (4 bytes long).

We didn't do this early on in adoption of Zarr version 3 because we didn't realize that Zarr version 3 doesn't detect the byte size of arrays being compressed with Blosc—unlike Zarr version 2 which does! So annoying...

hzorkic replied:
done !

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    token=None,
+    days_at_once=14,
+):
+    # Initialize Arraylake client
```
`sequential_reorg()` needs a docstring here describing what it does.

hzorkic replied:
done.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+        # If any day in the batch fails, exit the entire process,
+        # reporting the last successful day processed
+        except Exception as e:
+            print("\n   Batch failure:", e)
```
If you just re-raise the exception below, you don't need to include `e` in this output.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    return sfc_ds, pl_ds, inv_ds
+
+
+def sequential_reorg(
```
Function name is a little confusing. It looks like it is the primary entry point for the process of channelizing a dataset, so `channelize_dataset` maybe is a better name for this.

hzorkic replied:
done.

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    return in_locs, out_locs
+
+
+def process_day_from_datasets(
```
Choose a better name describing what it does. `channelize_day` maybe?

hzorkic replied:
done!

jeramey-rwe commented on 01_channelize_data/channelize.py:
```diff
@@ -0,0 +1,598 @@
+    pl_ds: xr.Dataset,
+):
+    """
+    Process exactly ONE calendar day and write its 4 timesteps.
```
A little more description of what this function should be doing here would be nice.

jeramey-rwe reviewed (changes requested):
Overall looks OK. Some suggested changes. `channelize.ipynb` appears to have some issues, but it's a bit hard to add review comments to that.

jeramey-rwe commented on 01_channelize_data/channelize.sh:
```diff
@@ -0,0 +1,77 @@
+DEST_BRANCH="extend-2025"
+DAYS_AT_ONCE=1
+
+MAX_RETRIES=100
```
This is a LOT of retries. Why are you doing so many retries?

hzorkic replied:
This is global retries. This script was handling daily 2015-2025 channelization of the ecmwf ICs. 10 years * 365 daty = 3650 opportunities to fail. At worst, the Network I/O failure occurs every 5-10 days, thus 100 retries. 

I decided on the brute force method here, so just retrying over and over again if the script failed for simplicity and then got annoyed having to restart the script. 99% of the time the error was the I/0 error and the other 1% was typos in the arguments I passed to the script.

jeramey-rwe replied:
Works for me!

hzorkic committed:
WIP still, but small Exception handling fixes

hzorkic committed:
update dataset_io to ensure we're not overwritting attrs

hzorkic committed:
updated channelize to prevent overwritting atrts

hzorkic committed:
update docstring for channelize_dataset

hzorkic committed:
add .sh script

hzorkic committed:
update gitignore

hzorkic committed:
refactor: huge changes to resample_era5.py
- remove day-batching logic so we can channelize larger chunks of data
- add ability to EXTEND existing dataset
- add metadata
- parallelize script
- extends capabilities to ERA5 and ECMWF

hzorkic committed:
update a small print statment and channelize notebook

hzorkic committed:
refactor: remove fodlers. move dataset_io.py to root. update gitignore.

hzorkic committed:
update valid_time metadata logic

hzorkic committed:
update .sh scripts arraylake_token errors
